README.md
# no-gather

Resources accumulate (trickle in) over time and women are champions.
Town and city phase upgrades are fast and free.

(a mod for [0ad](https://play0ad.com/))

The starting resource trickle rate is:

    54 wood
    54 food
    6  stone
    16 metal

every five seconds. When you reach P2, each number is multiplied by
1.3. When you reach P3, the multiplier is 1.5. Which means by P3, the
trickle rate is:

    105 wood
    105 wood
    11  stone
    31  metal

every 5 seconds.

Resources can not be gathered, you can only get them by allowing time
to pass and by using traders (and merchant ships if you're playing a
naval map).

When random maps are generated, they will have fewer trees and fewer
stone and metal mines.

## Issues and discussions

This can be done from the website at
https://github.com/0ad-matters/no-gather

There's also a [0ad forum
post](https://wildfiregames.com/forum/topic/92594-no-gather/) if you
prefer.


ChangeLog
2022-10-09

  * reduce default starting infantry to 0ad-public defaults
  * release v0.26.6

2022-09-13
  * release v0.26.5

2022-09-09

  * reduce the amounts of trees, food, metal and stone on random maps
  * significantly reduce stone and metal trickle

2022-09-08

  * Fix potential resource trickle OOS (thanks to Livv)
  * make town and city phase upgrades fast and free
  * release v0.26.3
  * resource rate increases with town and city phase upgrade
  * release v0.26.4

2022-09-07

  * fix starting unit counts
  * release v0.26.2
  * increase vision range of females (inherit)

2022-09-06

  * first release
  * disable storehouse, farmstead, corral, field
  * increase female attack
  * release v0.26.1



simulation/helpers/InitGame.js
/**
 * Called when the map has been loaded, but before the simulation has started.
 * Only called when a new game is started, not when loading a saved game.
 */
function PreInitGame()
{
	// no-gather mod
	const disabledTemplates = (civ) => [
		"structures/" + civ + "/storehouse",
		"structures/" + civ + "/field",
		"structures/" + civ + "/farmstead",
		"structures/" + civ + "/corral",
		];
	//

	// We need to replace skirmish "default" entities with real ones.
	// This needs to happen before AI initialization (in InitGame).
	// And we need to flush destroyed entities otherwise the AI gets the wrong game state in
	// the beginning and a bunch of "destroy" messages on turn 0, which just shouldn't happen.
	Engine.BroadcastMessage(MT_SkirmishReplace, {});
	Engine.FlushDestroyedEntities();

	let numPlayers = Engine.QueryInterface(SYSTEM_ENTITY, IID_PlayerManager).GetNumPlayers();
	for (let i = 1; i < numPlayers; ++i) // ignore gaia
	{
		let cmpTechnologyManager = QueryPlayerIDInterface(i, IID_TechnologyManager);
		if (cmpTechnologyManager)
			cmpTechnologyManager.UpdateAutoResearch();

		// no-gather mod
		let cmpPlayer = QueryPlayerIDInterface(i);
		const civ = QueryPlayerIDInterface(i, IID_Identity).GetCiv();
		cmpPlayer.SetDisabledTemplates(disabledTemplates(civ));
		//
	}

	// Explore the map inside the players' territory borders
	let cmpRangeManager = Engine.QueryInterface(SYSTEM_ENTITY, IID_RangeManager);
	cmpRangeManager.ExploreTerritories();
}

function InitGame(settings)
{
	// No settings when loading a map in Atlas, so do nothing
	if (!settings)
	{
		// Map dependent initialisations of components (i.e. garrisoned units)
		Engine.BroadcastMessage(MT_InitGame, {});
		return;
	}

	if (settings.ExploreMap)
	{
		let cmpRangeManager = Engine.QueryInterface(SYSTEM_ENTITY, IID_RangeManager);
		for (let i = 1; i < settings.PlayerData.length; ++i)
			cmpRangeManager.ExploreMap(i);
	}

	// Sandbox, Very Easy, Easy, Medium, Hard, Very Hard
	// rate apply on resource stockpiling as gathering and trading
	// time apply on building, upgrading, packing, training and technologies
	let rate = [ 0.42, 0.56, 0.75, 1.00, 1.25, 1.56 ];
	let time = [ 1.40, 1.25, 1.10, 1.00, 1.00, 1.00 ];
	let cmpModifiersManager = Engine.QueryInterface(SYSTEM_ENTITY, IID_ModifiersManager);
	let cmpAIManager = Engine.QueryInterface(SYSTEM_ENTITY, IID_AIManager);
	for (let i = 0; i < settings.PlayerData.length; ++i)
	{
		let cmpPlayer = QueryPlayerIDInterface(i);
		cmpPlayer.SetCheatsEnabled(!!settings.CheatsEnabled);

		if (settings.PlayerData[i] && !!settings.PlayerData[i].AI)
		{
			let AIDiff = +settings.PlayerData[i].AIDiff;
			cmpAIManager.AddPlayer(settings.PlayerData[i].AI, i, AIDiff, settings.PlayerData[i].AIBehavior || "random");
			cmpPlayer.SetAI(true);
			AIDiff = Math.min(AIDiff, rate.length - 1);
			cmpModifiersManager.AddModifiers("AI Bonus", {
				"ResourceGatherer/BaseSpeed": [{ "affects": ["Unit", "Structure"], "multiply": rate[AIDiff] }],
				"Trader/GainMultiplier": [{ "affects": ["Unit", "Structure"], "multiply": rate[AIDiff] }],
				"Cost/BuildTime": [{ "affects": ["Unit", "Structure"], "multiply": time[AIDiff] }],
			}, cmpPlayer.entity);
		}

		if (settings.PopulationCap)
			cmpPlayer.SetMaxPopulation(settings.PopulationCap);

		if (settings.AllyView)
			Engine.QueryInterface(cmpPlayer.entity, IID_TechnologyManager)?.ResearchTechnology(cmpPlayer.template.SharedLosTech);
	}
	if (settings.WorldPopulationCap)
		Engine.QueryInterface(SYSTEM_ENTITY, IID_PlayerManager).SetMaxWorldPopulation(settings.WorldPopulationCap);

	// Update the grid with all entities created for the map init.
	Engine.QueryInterface(SYSTEM_ENTITY, IID_Pathfinder).UpdateGrid();

	// Map or player data (handicap...) dependent initialisations of components (i.e. garrisoned units).
	Engine.BroadcastMessage(MT_InitGame, {});

	cmpAIManager.TryLoadSharedComponent();
	cmpAIManager.RunGamestateInit();
}

Engine.RegisterGlobal("PreInitGame", PreInitGame);
Engine.RegisterGlobal("InitGame", InitGame);


maps/random/rmgen2/gaia_no-gather.js
var treeReductionMultiplier = 0.3;

/**
 * Create huntable animals.
 */
function addAnimals(constraint, size, deviation, fill)
{
	return;
}

function addBerries(constraint, size, deviation, fill)
{
	return;
}

function addFish(constraint, size, deviation, fill)
{
	return;
}

function addForests(constraint, size, deviation, fill)
{
	if (currentBiome() == "generic/savanna")
		return;

	g_Map.log("Creating forests");

	let treeTypes = [
		[
			g_Terrains.forestFloor2 + TERRAIN_SEPARATOR + g_Gaia.tree1,
			g_Terrains.forestFloor2 + TERRAIN_SEPARATOR + g_Gaia.tree2,
			g_Terrains.forestFloor2
		],
		[
			g_Terrains.forestFloor1 + TERRAIN_SEPARATOR + g_Gaia.tree4,
			g_Terrains.forestFloor1 + TERRAIN_SEPARATOR + g_Gaia.tree5,
			g_Terrains.forestFloor1
		]
	];

	let forestTypes = [
		[
			[g_Terrains.forestFloor2, g_Terrains.mainTerrain, treeTypes[0]],
			[g_Terrains.forestFloor2, treeTypes[0]]
		],
		[
			[g_Terrains.forestFloor2, g_Terrains.mainTerrain, treeTypes[1]],
			[g_Terrains.forestFloor1, treeTypes[1]]],
		[
			[g_Terrains.forestFloor1, g_Terrains.mainTerrain, treeTypes[0]],
			[g_Terrains.forestFloor2, treeTypes[0]]],
		[
			[g_Terrains.forestFloor1, g_Terrains.mainTerrain, treeTypes[1]],
			[g_Terrains.forestFloor1, treeTypes[1]]
		]
	];

	for (let forestType of forestTypes)
	{
		let offset = getRandomDeviation(size, deviation);
		createAreas(
			new ChainPlacer(1, Math.floor(scaleByMapSize(3, 5) * offset), Math.floor(50 * offset), 0.5),
			[
				new LayeredPainter(forestType, [2]),
				new TileClassPainter(g_TileClasses.forest)
			],
			constraint,
			10 * fill * treeReductionMultiplier);
	}
}

function addMetal(constraint, size, deviation, fill)
{
	return;
}

/**
 * Create stone mines.
 */
function addStone(constraint, size, deviation, fill)
{
	return;
}

/**
 * Create straggler trees.
 */
function addStragglerTrees(constraint, size, deviation, fill)
{
	g_Map.log("Creating straggler trees");

	// Ensure minimum distribution on african biome
	if (currentBiome() == "generic/savanna")
	{
		fill = Math.max(fill, 2);
		size = Math.max(size, 1);
	}

	var trees = [g_Gaia.tree1, g_Gaia.tree2, g_Gaia.tree3, g_Gaia.tree4];

	var treesPerPlayer = 40;
	var playerBonus = Math.max(1, (getNumPlayers() - 3) / 2);

	var offset = getRandomDeviation(size, deviation);
	var treeCount = treesPerPlayer * playerBonus * fill * treeReductionMultiplier;
	var totalTrees = scaleByMapSize(treeCount, treeCount);

	var count = Math.floor(totalTrees / trees.length) * fill;
	var min = offset;
	var max = 4 * offset;
	var minDist = offset;
	var maxDist = 5 * offset;

	// More trees for the african biome
	if (currentBiome() == "generic/savanna")
	{
		min = 3 * offset;
		max = 5 * offset;
		minDist = 2 * offset + 1;
		maxDist = 3 * offset + 2;
	}

	for (var i = 0; i < trees.length; ++i)
	{
		var treesMax = max;

		// Don't clump fruit trees
		if (i == 2 && (currentBiome() == "generic/sahara" || currentBiome() == "generic/aegean"))
			treesMax = 1;

		min = Math.min(min, treesMax);

		var group = new SimpleGroup([new SimpleObject(trees[i], min, treesMax, minDist, maxDist)], true, g_TileClasses.forest);
		createObjectGroupsDeprecated(group, 0, constraint, count);
	}
}


maps/random/rmgen-common/player_no-gather.js
/**
 * @file These functions locate and place the starting entities of players.
 */

/**
 * These are identifiers of functions that can generate parts of a player base.
 * There must be a function starting with placePlayerBase and ending with this name.
 * This is a global so mods can extend this from external files.
 */
var g_PlayerBaseFunctions = [
	// Possibly mark player class first here and use it afterwards
	"CityPatch",
	// Create the largest and most important entities first
	"Treasures",
	"Decoratives"
];

/**
 * Places the given entities at the given location (typically a civic center and starting units).
 * @param location - A Vector2D specifying tile coordinates.
 * @param civEntities - An array of objects with the Template property and optionally a Count property.
 * The first entity is placed in the center, the other ones surround it.
 */
function placeStartingEntities(location, playerID, civEntities, dist = 6, orientation = BUILDING_ORIENTATION)
{
	// Place the central structure
	let i = 0;
	let firstTemplate = civEntities[i].Template;
	if (firstTemplate.startsWith("structures/"))
	{
		g_Map.placeEntityPassable(firstTemplate, playerID, location, orientation);
		++i;
	}

	// Place entities surrounding it
	let space = 2;
	for (let j = i; j < civEntities.length; ++j)
	{
		let angle = orientation - Math.PI * (1 - j / 2);
		let count = civEntities[j].Count || 1;

		for (let num = 0; num < count; ++num)
		{
			let position = Vector2D.sum([
				location,
				new Vector2D(dist, 0).rotate(-angle),
				new Vector2D(space * (-num + (count - 1) / 2), 0).rotate(angle)
			]);

			g_Map.placeEntityPassable(civEntities[j].Template, playerID, position, angle);
		}
	}
}


maps/random/rmgen-common/gaia_entities_no-gather.js
var treeReductionMultiplier = 0.3

/**
 * @file These functions are often used to place gaia entities, like forests, mines, animals or decorative bushes.
 */

/**
 * Returns the number of trees in forests and straggler trees.
 */
function getTreeCounts(minTrees, maxTrees, forestRatio)
{
	return [forestRatio, 1 - forestRatio].map(p => p * scaleByMapSize(minTrees * treeReductionMultiplier, maxTrees * treeReductionMultiplier));
}

function createStragglerTrees(templateNames, constraint, tileClass, treeCount, retryFactor)
{
	g_Map.log("Creating straggler trees");
	for (let templateName of templateNames)
		createObjectGroupsDeprecated(
			new SimpleGroup([new SimpleObject(templateName, 1, 1, 0, 3)], true, tileClass),
			0,
			constraint,
			Math.floor(treeCount * treeReductionMultiplier / templateNames.length),
			retryFactor);
}

function createMines(objects, constraint, tileClass, count)
{
	return;
}

function createBalancedMines(oSmall, oLarge, clMine, constraints, counts, randomness)
{
	return;
}

function createBalancedMetalMines(oSmall, oLarge, clMine, constraints, counts = 1.0, randomness = 0.05)
{
	return;
}

function createBalancedStoneMines(oSmall, oLarge, clMine, constraints, counts = 1.0, randomness = 0.05)
{
	return;
}

function createStoneMineFormation(position, templateName, terrain, radius = 2.5, count = 8, startAngle = undefined, maxOffset = 1)
{
	return;
}

function createFood(objects, counts, constraint, tileClass)
{
	return;
}


maps/random/rmgen-common/gaia_entities.js
/**
 * @file These functions are often used to place gaia entities, like forests, mines, animals or decorative bushes.
 */

/**
 * Returns the number of trees in forests and straggler trees.
 */
function getTreeCounts(minTrees, maxTrees, forestRatio) // no-gather mod
{
	return [forestRatio, 1 - forestRatio].map(p => p * scaleByMapSize(minTrees * 0.6, maxTrees * 0.6));
}

/**
 * Places uniformly sized forests at random locations.
 * Unless you want a custom number of forest, prefer createDefaultForests.
 * Generates two variants of forests from the given terrain textures and tree templates.
 * The forest border has less trees than the inside.
 * @param terrainsSet - a list of 5 terrains to use. The first 3 are border terrains, the later 2 interior.
 * @param constraint - constraints to respect
 * @param tileClass - the tileclass to print
 * @param treeCount - Either { "nbForests": X, "treesPerForest": X } or (legacy) a number of trees.
 * @param retryFactor - @see createAreas
 */
function createForests(terrainSet, constraint, tileClass, treeCount, retryFactor)
{
	if (!treeCount)
		return;

	// Construct different forest types from the terrain textures and template names.
	const [mainTerrain, terrainForestFloor1, terrainForestFloor2, terrainForestTree1, terrainForestTree2] = terrainSet;

	// The painter will pick a random Terrain for each part of the forest.
	const forestVariants = [
		{
			"borderTerrains": [terrainForestFloor2, mainTerrain, terrainForestTree1],
			"interiorTerrains": [terrainForestFloor2, terrainForestTree1]
		},
		{
			"borderTerrains": [terrainForestFloor1, mainTerrain, terrainForestTree2],
			"interiorTerrains": [terrainForestFloor1, terrainForestTree2]
		}
	];

	let numberOfTrees;
	let numberOfForests;
	if (typeof treeCount === "number")
	{
		numberOfTrees = treeCount;
		numberOfForests = Math.floor(numberOfTrees / (scaleByMapSize(3, 6) * getNumPlayers() * forestVariants.length));
	}
	else
	{
		numberOfForests = treeCount.nbForests;
		numberOfTrees = numberOfForests * treeCount.treesPerForest;
	}

	if (!numberOfForests)
		return;

	g_Map.log("Creating forests");
	for (const forestVariant of forestVariants)
		createAreas(
			new ChainPlacer(1, Math.floor(scaleByMapSize(3, 5)), numberOfTrees / numberOfForests * 0.2, 0.5),
			[
				new LayeredPainter([forestVariant.borderTerrains, forestVariant.interiorTerrains], [2]),
				new TileClassPainter(tileClass)
			],
			constraint,
			numberOfForests * 0.7,
			retryFactor);
}

var g_DefaultNumberOfForests = scaleByMapSize(8, 36);

/**
 * Passes some sane defaults to createForests.
 */
function createDefaultForests(terrainSet, constraints, tileClass, totalNumberOfTrees)
{
	return createForests(
		terrainSet,
		constraints,
		tileClass,
		{
			"nbForests": g_DefaultNumberOfForests,
			"treesPerForest": totalNumberOfTrees / g_DefaultNumberOfForests,
		}
	);
}


/**
 * Places the given amount of Entities at random places meeting the given Constraint, chosing a different template for each.
 */
function createStragglerTrees(templateNames, constraint, tileClass, treeCount, retryFactor)
{
	g_Map.log("Creating straggler trees");
	for (let templateName of templateNames)
		createObjectGroupsDeprecated(
			new SimpleGroup([new SimpleObject(templateName, 1, 1, 0, 3)], true, tileClass),
			0,
			constraint,
			Math.floor(treeCount * 0.6 / templateNames.length), // no-gather mod
			retryFactor);
}

/**
 * Places a SimpleGroup consisting of the given number of the given Objects
 * at random locations that meet the given Constraint.
 */
function createMines(objects, constraint, tileClass, count)
{
	return; // no-gather mod
	for (let object of objects)
		createObjectGroupsDeprecated(
			new SimpleGroup(object, true, tileClass),
			0,
			constraint,
			count || scaleByMapSize(4, 16),
			70);
}

/**
 * Place large/small mines on the map in such a way that it should be relatively fair.
 * @param oSmall - the small mine object
 * @param oLarge - the large mine object
 * @param clMine - the 'mine' class to paint.
 * @param constraints - Custom constraints. Note that the function automatically avoids clMine as well.
 * @param counts - a dict of numbers { "largeCount": 10, "smallCount": 100, "randomSmallCount": 2 }
 * @param randomness - randomize counts by a random multiplier between [1 - randomness, 1 + randomness]
 */
function createBalancedMines(oSmall, oLarge, clMine, constraints, counts, randomness)
{
	// no-gather mod
	return;
	let largeCount = counts.largeCount;
	let smallCount = counts.smallCount;
	let randomSmallCount = counts.randomSmallCount;

	if (randomness > 0 && randomness < 1)
	{
		largeCount = Math.round(largeCount * randFloat(1 - randomness, 1 + randomness));
		smallCount = Math.round(smallCount * randFloat(1 - randomness, 1 + randomness));
		randomSmallCount = Math.round(randomSmallCount * randFloat(1 - randomness, 1 + randomness));
	}

	const arrayConstraints = Array.isArray(constraints) ? constraints : [constraints];

	// Plop large mines far away from each other.
	createObjectGroups(
		new SimpleGroup([new SimpleObject(oLarge, 1, 1, 0, 1)], true, clMine),
		0,
		new AndConstraint([avoidClasses(clMine, scaleByMapSize(25, 50)), ...arrayConstraints]),
		largeCount,
		100);

	// Plop smaller clusters of small mines, also somewhat farther away.
	createObjectGroups(
		new SimpleGroup([new SimpleObject(oSmall, 2, 3, 0, 2)], true, clMine),
		0,
		new AndConstraint([avoidClasses(clMine, scaleByMapSize(18, 35)), ...arrayConstraints]),
		smallCount,
		50);

	// Plop a few smaller clusters in a random fashion, occasionally making very good dropsites spots.
	createObjectGroups(
		new SimpleGroup([new SimpleObject(oSmall, 1, 2, 0, 2)], true, clMine),
		0,
		new AndConstraint([avoidClasses(clMine, 5), ...arrayConstraints]),
		randomSmallCount,
		50);
}

/**
 * Helper for createBalancedMines with default metal counts.
 * The current settings are so that a Small 1v1 has about 40K metal,
 * and a Normal 4v4 has about 140K.
 * The setup is biaised so that with fewer players, there are more small mines,
 * and with more players there are proportionally more big mines, to maintain
 * some randomness to the distribution but keep it somewhat fair in 1v1.
 */
function createBalancedMetalMines(oSmall, oLarge, clMine, constraints, counts = 1.0, randomness = 0.05)
{
	return createBalancedMines(
		oSmall,
		oLarge,
		clMine,
		constraints,
		{
			"largeCount": (Math.max(scaleByMapSize(1, 9), getNumPlayers() * 1.8 - 0.8)) * counts,
			"smallCount": (scaleByMapSize(4, 12)) * counts,
			"randomSmallCount": (scaleByMapSize(1, 8)) * counts,
		},
		randomness
	);
}

/**
 * Helper for createBalancedMines with default stone counts.
 * There is a little less stone than metal overall.
 */
function createBalancedStoneMines(oSmall, oLarge, clMine, constraints, counts = 1.0, randomness = 0.05)
{
	return createBalancedMines(
		oSmall,
		oLarge,
		clMine,
		constraints,
		{
			"largeCount": (Math.max(scaleByMapSize(1, 9), getNumPlayers() * 1.25)) * counts,
			"smallCount": (scaleByMapSize(1, 8)) * counts,
			"randomSmallCount": (scaleByMapSize(1, 8)) * counts,
		},
		randomness
	);
}


/**
 * Places Entities of the given templateName in a circular pattern (leaving out a quarter of the circle).
 */
function createStoneMineFormation(position, templateName, terrain, radius = 2.5, count = 8, startAngle = undefined, maxOffset = 1)
{
	createArea(
		new ChainPlacer(radius / 2, radius, 2, Infinity, position, undefined, [5]),
		new TerrainPainter(terrain));

	let angle = startAngle !== undefined ? startAngle : randomAngle();

	for (let i = 0; i < count; ++i)
	{
		let pos = Vector2D.add(position, new Vector2D(radius + randFloat(0, maxOffset), 0).rotate(-angle)).round();
		g_Map.placeEntityPassable(templateName, 0, pos, randomAngle());
		angle += 3/2 * Math.PI / count;
	}
}

/**
 * Places the given amounts of the given Objects at random locations meeting the given Constraint.
 */
function createFood(objects, counts, constraint, tileClass)
{
	return; // no-gather mod
	g_Map.log("Creating food");
	for (let i = 0; i < objects.length; ++i)
		createObjectGroupsDeprecated(
			new SimpleGroup(objects[i], true, tileClass),
			0,
			constraint,
			counts[i],
			50);
}

/**
 * Same as createFood, but doesn't mark the terrain with a TileClass.
 */
function createDecoration(objects, counts, constraint)
{
	g_Map.log("Creating decoration");
	for (let i = 0; i < objects.length; ++i)
		createObjectGroupsDeprecated(
			new SimpleGroup(objects[i], true),
			0,
			constraint,
			counts[i],
			5);
}

/**
 * Places docks in situations where the location of land and water is not known in advance.
 * Do determine the position, it picks a random point on the land, find the closest, significantly large body of water,
 * then places the dock at the first point close to that body of water within the given heightrange.
 *
 * @param {string} template - The template name of the dock to be placed.
 * @param {number} playerID - The owner of the dock.
 * @param {number} count - The number of docks to be placed.
 * @param {Object} tileClassWater - The tileclass the water area is marked with.
 * @param {Object} tileClassDock - The dock position is marked with this class.
 * @param {number} heightMin - The lowest height a dock could be placed.
 * @param {number} heightMax - The greatest height a dock could be placed.
 * @param {Array|Constraint} constraints - Only consider dock positions valid that meet this Constraint.
 * @param {number} offset - How many tiles to move the dock towards the direction of the water after having found a location.
 * @param {number} retryFactor- How many different locations should be tested.
 */
function placeDocks(template, playerID, count, tileClassWater, tileClassDock, heightMin, heightMax, constraints, offset, retryFactor)
{
	let mapCenter = g_Map.getCenter();

	g_Map.log("Marking dock search start area");
	let areaSearchStart = createArea(
		new DiskPlacer(fractionToTiles(0.5) - 10, mapCenter),
		undefined,
		avoidClasses(tileClassWater, 6));

	g_Map.log("Marking dock search end area");
	let areaSearchEnd = createArea(
		new DiskPlacer(fractionToTiles(0.5) - 10, mapCenter),
		undefined,
		stayClasses(tileClassWater, 20));

	g_Map.log("Marking land area");
	let areaLand = createArea(
		new MapBoundsPlacer(),
		undefined,
		avoidClasses(tileClassWater, 0));

	g_Map.log("Marking water area");
	let areaWater = createArea(
		new MapBoundsPlacer(),
		undefined,
		stayClasses(tileClassWater, 0));

	if (!areaSearchEnd || !areaSearchEnd.getPoints().length)
		return;

	// TODO: computing the exact intersection with the waterplane would both not require us to pass reasonable heights and be more precise
	let constraint = new AndConstraint(constraints);
	g_Map.log("Placing docks");
	for (let i = 0; i < count; ++i)
		for (let tries = 0; tries < retryFactor; ++tries)
		{
			let positionLand = pickRandom(areaSearchStart.getPoints());
			let positionWaterLarge = areaSearchEnd.getClosestPointTo(positionLand);
			let positionDock = findLocationInDirectionBasedOnHeight(positionWaterLarge, positionLand, heightMin, heightMax, offset);
			if (!positionDock)
				continue;

			positionDock.round();

			if (!g_Map.inMapBounds(positionDock) || !constraint.allows(positionDock))
				continue;

			let angle = positionDock.angleTo(Vector2D.average(new DiskPlacer(8, positionDock).place(stayClasses(tileClassWater, 0))));

			g_Map.placeEntityPassable(template, playerID, positionDock, -angle + Math.PI / 2);
			tileClassDock.add(positionDock);
			break;
		}
}
